src/main/java/com/google/sps/servlets/AuthenticationServlet.java

```diff
@@ -0,0 +1,122 @@
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.gson.Gson;
+import java.util.HashMap;
+import java.util.Map;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+
+/**
+ * The authentication servlet is responsible for authenticating users and retreiving user
+ * preferences from datastore.
+ *
+ * @author  Aidan Molloy
+ */
+@WebServlet("/auth")
+public class AuthenticationServlet extends HttpServlet {
+  /**
+   * Checks if the user is logged in with UserService and prints response in json.
+   *
+   * @param   request     provides request information from the HTTP servlet
+   * @param   response    response object where servlet will write information to
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json;");
+    UserService userService = UserServiceFactory.getUserService();
+    Map<String, String> authResponse = new HashMap<String, String>();
+
+    try{
+      if (userService.isUserLoggedIn()) {
+        // If logged in get email and create link to logout
+        String userEmail = userService.getCurrentUser().getEmail();
+        String logoutUrl = userService.createLogoutURL("/");
+
+        authResponse.put("email", userEmail);
+        authResponse.put("logoutUrl", logoutUrl);
+
+        // Get User Info and add it to the response
+        authResponse.putAll(getUserInfo(userService.getCurrentUser().getEmail()));
+      } else {
+        // If logged out get login link
+        String loginUrl = userService.createLoginURL("/");
+        authResponse.put("loginUrl", loginUrl);
+      }
+    }catch(Exception e) {
+      authResponse.put("errorMsg", "Something went wrong. Please try again later.");
+    }
+
+    String json = convertToJson(authResponse);
+
+    // Write response to /auth
+    response.getWriter().println(json);
+  }
+
+  /**
+   * Returns the preferences of the user with email, or null if the user has not set their 
+   * preferences.
+   *
+   * @param   email   the email of logged in user to retreive linked userInfo from datastore
+   * @return          a map of userInfo linked to the user email
+   */
+  private Map<String, String> getUserInfo(String email) {
+    Map<String, String> userInfoResponse = new HashMap<String, String>();
+    try{
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      Query query =
+          new Query("UserInfo")
+              .setFilter(new Query.FilterPredicate("email", Query.FilterOperator.EQUAL, email));
```

Comment:
Is email guaranteed to be unique per user?

Reply:
This is what we decided to go with as the unique identifier between users, for university level lecturers generally identify their students by their student emails.